# Introduction to Numpy Datetime

## Learning objectives 

By the end of this unit you should be able to:
- Use the `numpy` classes and functions for handling dates and times.
- Understand the difference between `numpy` and Python versions of datetime handling.
- Use `numpy.datetime64` inside more general `numpy` functions.
- Handle arithmetic with `numpy.datetime64` using rescaling.

## What is `numpy.datetime64`?

- A class for handling dates and times built into the `numpy` library for scientific computing in Python.

- An alternative to the Python `datetime` class with improved memory efficiency and easy integration with `numpy` functions.

- Use `datetime64` objects in lists and arrays just like number types such as `np.float`. The `64` in the name refers to `numpy` storing each object as `64 bit`  integers.

In [34]:
# We often import numpy library with the alias "np"
import numpy as np
print(np.datetime64)

<class 'numpy.datetime64'>


The Numpy library is a tool for working with any numbers in Python. This package brings fast and memory efficient handling of arrays and matrices to Python that isn't possible using only native `List` objects. Numpy promotes numerical precision and array-focused optimisations to your code. Numpy is one of the most fundamental tools for data science in Python. Without Numpy, the Python language might never have become the most popular programming language for data science or machine learning.

The core parts of Numpy are focused on optimising operations for numbers but there is also the facility to handle datetime objects. We previously looked at the Python `datetime` class and now we'll investigate the Numpy `datetime64` class. This class is also focused on handling dates and times, with many similar behaviours and interfaces to the Python `datetime` package we are familiar with. In this lesson, we will learn about the `datetime64` class, compare it to the Python `datetime` class and learn how to cast between the two. We will also explore some unique features of Numpy `datetime64` and how to use them. To get the most out of this tutorial, make sure you've completed the previous lesson on Python datetime.

## Creating `np.datetime64` objects.

To declare a `datetime64` object you can either pass an **ISO 8601** formatted string:

In [35]:
from datetime import datetime
import dateutil.tz as tz
import pandas as pd

In [36]:
dt1 = np.datetime64("2018-10-03 13:00")
print(dt1)

2018-10-03T13:00


Or initialise using a Python `datetime` object:

In [37]:
pydt1 = datetime(year=2018, month=10, day=3, hour=13)
dt2 = np.datetime64(pydt1)
print(dt2)

2018-10-03T13:00:00.000000


We declare a Numpy `datetime64` object using a single argument. The `np.datetime64` class takes either a string input or an existing Python `datetime` object as input. This string **must** be formatted in the ISO 8601 specification, where the largest unit (year) comes first and the smallest unit (seconds) comes last. You can also initialise a `numpy` object from an existing Python `datetime` object. ([documentation](https://www.iso.org/iso-8601-date-and-time-format.html))

But you **cannot** declare using individual date and time features as arguments, ie using the same declaration pattern as for Python `datetime`, as can be seen in the example below:

```python
dt3 = np.datetime64(year=2018, month=10, day=3, hour=13)
```

## Comparing `np.datetime64` to `datetime`

Both classes will store the information of a date and time. But the `numpy` version is more **memory efficient**. 

In [38]:
import sys

pdt1 = datetime(year=2018, month=10, day=3, hour=13)
ndt1 = np.datetime64("2018-10-03 13:00:00")

# Compare size in bytes
print(sys.getsizeof(pdt1))
print(sys.getsizeof(ndt1)) 

48
40


When we previously studied the Python `datetime` package we looked at all the different features of the package and how the class `datetime` fits into the package `datetime`. Now we will look at the `datetime64` class slightly differently, as this is a class that fits into a wider Numpy ecosystem. Not everything we learnt for Python `datetime` will be applicable here, as well as some features for Numpy don't exist for basic Python. The Numpy `datetime64` class is not always a direct replacement and learning about this class doesn't mean you should discard the Python `datetime` class. Each class will be useful for different occasions and you will likely find yourself transferring between the two.  

Now that we can declare Numpy objects, we can compare between Numpy and Python versions of a datetime object. The Numpy library is optimised for memory and efficiency. We can check how large an object is in `bytes` using the `getsizeof` function from the built-in `sys` module. Comparing between two objects, _which hold the same information_ , we see that the Numpy object saves 8 bytes compared to native Python. This might not seem like much now. But what if you start to have massive arrays of thousands of dates?

## What about sequences of datetime objects?

In [39]:
pdt2 = datetime(year=2019, month=10, day=4, hour=2)
pdt3 = datetime(year=2020, month=1, day=3, hour=6)
native_array = [pdt1, pdt2, pdt3]
print(sys.getsizeof(native_array))

ndt2 = np.datetime64("2019-10-04 12:46:00")
ndt3 = np.datetime64("2020-01-03 18:00:32")
numpy_array = np.array([ndt1, ndt2, ndt3], dtype='datetime64')
print(sys.getsizeof(numpy_array))

80
136


Python's `datetime` objects are high-level abstractions, and storing them in a list introduces overhead due to their object-oriented nature and additional metadata. This becomes noticeable when handling even a small number of `datetime` objects.

On the other hand, NumPy, a library optimized for numerical computations, offers the `np.datetime64` class, which is designed for efficient storage and manipulation of date and time data. NumPy arrays store these objects in a compact, low-level format, making them more memory-efficient when working with larger datasets.

In the example provided, the memory usage of a native Python list containing `datetime` objects is compared to a NumPy array of `datetime64` objects. While the NumPy array demonstrates superior memory efficiency for larger arrays, the benefits may not be as apparent for smaller arrays due to NumPy's memory management overhead.

**In summary**, NumPy is highly advantageous for handling large arrays of datetime data, offering significant memory savings and computational efficiency. However, for smaller arrays, the memory optimization may be negligible.

## Converting between `np.datetime64` and `datetime`

Converting between Python `datetime` and Numpy `datetime64` is typically easy. We saw before that you can initialise a Numpy object from a Python object. This is exactly how we can create Numpy objects out of existing Python objects in one line. To reverse this procedure and get a Python object from a Numpy object, we need to call the `np.astype` function on the `Numpy` object. 

In [40]:
pdt_now = datetime.now() # datetime in Python
print(pdt_now)

ndt_now = np.datetime64(pdt_now) # Convert to Numpy
print(ndt_now)

pdt_fromnp = ndt_now.astype(datetime) # Convert Numpy back to datetime
print(pdt_fromnp)
print(pdt_now == pdt_fromnp)

2026-08-16 17:44:45.455570
2026-08-16T17:44:45.455570
2026-08-16 17:44:45.455570
True


However, there is one important **caveat** to converting between Python and Numpy: Numpy does not respect or preserve **aware** datetimes, meaning datetimes with an attached timezone. Always use naive datetimes in Numpy.

## Using `np.datetime64` in `numpy `

Numpy can generate sequences of datetime objects using the standard Numpy library.

In [41]:
# Use arange to extract every day in February 2020
np.arange("2020-02-01", "2020-03-01", dtype="datetime64")

array(['2020-02-01', '2020-02-02', '2020-02-03', '2020-02-04',
       '2020-02-05', '2020-02-06', '2020-02-07', '2020-02-08',
       '2020-02-09', '2020-02-10', '2020-02-11', '2020-02-12',
       '2020-02-13', '2020-02-14', '2020-02-15', '2020-02-16',
       '2020-02-17', '2020-02-18', '2020-02-19', '2020-02-20',
       '2020-02-21', '2020-02-22', '2020-02-23', '2020-02-24',
       '2020-02-25', '2020-02-26', '2020-02-27', '2020-02-28',
       '2020-02-29'], dtype='datetime64[D]')

In [42]:
# Using the step argument to return every other day
np.arange("2020-02-01", "2020-03-01", step=2, dtype="datetime64")

array(['2020-02-01', '2020-02-03', '2020-02-05', '2020-02-07',
       '2020-02-09', '2020-02-11', '2020-02-13', '2020-02-15',
       '2020-02-17', '2020-02-19', '2020-02-21', '2020-02-23',
       '2020-02-25', '2020-02-27', '2020-02-29'], dtype='datetime64[D]')

**What is this `dtype='datetime64[D]'`?**

When comparing memory footprint between Python and Numpy we declared a standard `np.array` object with elements of type `np.datetime64` without explanation. We'll now discuss this further to unpack how Numpy makes sequences of datetime objects manageable. Typically, when considering Numpy arrays we use integer or float numbers and imagine them as big N-dimensional data structures full of a basic (but precise) numbers. Conversely, when looking at datetime objects, we examine how they are anything _but_ basic with many functions and utilities associated with the class. Numpy brings these two features together to allow you to use complex `np.datetime64` objects inside Numpy arrays in **exactly the same fashion as you would for number types**. This is why we can declare a `np.array()` object, specify the element class as `np.datetime64` (using `dtype=np.datetime64`) and then manipulate this data with efficient Numpy operations.

Because `datetime64` behaves just like a number in the Numpy library, we can use functions such as `np.arange()` to generate sequences of datetimes. For `np.arange()` we specify the start datetime (inclusive) and the end datetime (exclusive) as ISO8601 strings. We can also use the `step` argument to extract every _other_ day. To decide on the increment, Numpy will use the smallest specified unit. Since we provided only dates without times, Numpy increased the *day* in each element. We'll discuss how to modify this later but as a hint, the key is in the `[D]` part of the `array.dtype` attribute.

## Resolution and Rescaling

**What does `dtype='datetime64[D]'` mean?** 


Numpy uses a set of **time units** to specify ranges and the **resolution** of a `datetime64` object. 

Specifying the time unit in Numpy functions on `datetime64` objects controls how Numpy handles datetimes manipulation.

The full specification can be found in the Numpy docs [Numpy Datetime Unit Documentation](https://numpy.org/doc/stable/reference/arrays.datetime.html?highlight=datetime#datetime-units)

In [43]:
df = pd.DataFrame({"Code": ["Y", "M", "W", "D", "h", "m", "s", "ms", "..."], 
                   "Time Unit": ["year", "month", "week", "day", "hour", "minute", "second", "millisecond", "smaller units..."]})

df

,Code,Time Unit
0,Y,year
1,M,month
2,W,week
3,D,day
4,h,hour
5,m,minute
6,s,second
7,ms,millisecond
8,...,smaller units...


The `[D]` part of the Numpy arrays we just created indicated that those arrays were incrementing in Days. Numpy has a set codes to indicate the unit of time that is being varied or changed. This is different to how Python indexes time units and we will see how this approach can be extremely useful in arithmetic. We show the codes for Year to millisecond but you can specify down to attosecond (one quintillionth of a second) if required. We will also see how these units can be used with raw numbers to specify changes in arithmetic operations. Using a specific unit will change how Numpy manipulates data and can make handling datetime objects far easier than possible in Python alone.

## Examples using Datetime units

In [44]:
# Use the second argument to datetime64 to specify precision of the object
np.datetime64("2020-01-02", "ns")

numpy.datetime64('2020-01-02T00:00:00.000000000')

In [45]:
# Specifying hours as the smallest unit - Numpy will increment hours
dt = np.arange("2020-02-01T00", "2020-03-01T00", step=2, dtype="datetime64")
print(dt[:8])

['2020-02-01T00' '2020-02-01T02' '2020-02-01T04' '2020-02-01T06'
 '2020-02-01T08' '2020-02-01T10' '2020-02-01T12' '2020-02-01T14']


We can use these time codes to ensure **consistent** behaviour when handling datetime64 objects. The `np.datetime64` constructor can take a maximum of two arguments with the second argument, a time unit code, specifying the **precision** of the object. We've declared an object with only year, month and day but specified `ns` (nanoseconds) in the second argument. This forces Numpy to fill in the object with values down to this precision. Conversely, if we change `ns` to `Y` (year), then Numpy would only retain the value of Year and drop the smaller datetime details.

We can also look at how these time units work for **varying** datetime objects in the `np.arange()` function. First, we specify hours as the smallest unit and let Numpy guess the unit to increment. Whichever unit that `numpy` decides to vary will impact whether the array spans minutes, hours, days or any other time unit. Here we have specified to select every **other** unit using the `step` argument. The end result is an array over `hours`, increasing by 2 in each element.

In [46]:
# Or use datetime unit code to achieve the same result
dt = np.arange("2020-02-01", "2020-03-01", step=2, dtype="datetime64[h]")
print(dt[:8])

['2020-02-01T00' '2020-02-01T02' '2020-02-01T04' '2020-02-01T06'
 '2020-02-01T08' '2020-02-01T10' '2020-02-01T12' '2020-02-01T14']


In [47]:
# Numpy increments month since we specify [M]
dt = np.arange("2020-02-05", "2020-08-07", dtype="datetime64[M]")
print(dt)

['2020-02' '2020-03' '2020-04' '2020-05' '2020-06' '2020-07']


The preferred approach to ensuring that Numpy uses hours is by giving the `[h]` code to `dtype`. While these two approaches give the same result, the second strategy using time unit codes removes any ambiguity. For example, if you are importing data formatted with hours, minutes and seconds but you need to increment hours between each imported item -- it would be far easier to use `dtype=datetime64[h]` in `np.arange()` than to splice each date so that Numpy guesses to increment hours. We can also increment larger time units such as month using `[M]`, but it is important to remember that Numpy will "cut off" the smaller time units. Here, we see that Numpy cuts off days when asking for the months between February and August. 

## Useful functions

* Numpy includes a `timedelta64` class, similar to the Python `datetime.timedelta`, for time **intervals** and **differences** between datetime objects. 
* We can use the **time units** to specify the length of the `timedelta64`.
* Numpy supports only one unit in a `timedelta64` object.

In [48]:
np.timedelta64(14, "m") # 14 minutes

numpy.timedelta64(14,'m')

In [49]:
np.timedelta64(366, "ms") # 366 milliseconds

numpy.timedelta64(366,'ms')

Numpy also has a class for time intervals similar to Python. Before, the Python `timedelta` class accepted arguments for each time unit. However, for Numpy we need to use the codes we just saw to specify the duration of the delta. In addition, Numpy supports only one unit for a `timedelta64` object. The same types of units from the ISO8601 format are accepted so here we have declared `timedelta64` objects in minutes (`m`) and milliseconds (`ms`). Note that these units are case sensitive - a capital `M` will give you a `timedelta64` in months not minutes!

In [50]:
np.timedelta64(18, "W") # 18 weeks

numpy.timedelta64(18,'W')

In [51]:
np.timedelta64(7, 'h') + np.timedelta64(7, 'm') # 7 hours and 7 minutes

numpy.timedelta64(427,'m')

Similar to Python, the `numpy` class supports weeks as `W`. In our last example, we desired an interval of 7 hours and 7 minutes. To support this, all units are converted to the smallest unit required. Numpy stores the output as 427 minutes, which is equivalent to the desired duration in a single unit. 

Convert a `np.timedelta64` interval into a new interval by **creating a new object** with the desired time unit.

In [52]:
dt = np.timedelta64(7, 'h') + np.timedelta64(7, 'm') + np.timedelta64(366, "ms")
np.timedelta64(dt, "s") # Create a new timedelta64 in seconds

numpy.timedelta64(25620,'s')

Or extract the whole seconds in a `timedelta64` object using `np.astype()`.

In [53]:
dt.astype("timedelta64[s]") # Convert to seconds but drop the 366ms

numpy.timedelta64(25620,'s')

To change the unit of a `timedelta64` object, we can create a new object with the desired unit code. We can also use casting with `np.astype()` to convert timedelta objects into different units. These two approaches are mostly similar but with a subtle distinction. Creating a new object will check if the conversion is valid (more on this shortly) whereas using `np.astype()` will convert the duration using default values. These two will give you the same result until you need to convert time units with inconsistent lengths.

Here we create a timedelta containing hours, minutes and milliseconds. We want to determine how many seconds are in this interval. The fastest approach is to cast the object to `timedelta64[s]`, which transforms the object into seconds. However, this will drop the units smaller than a second meaning we lose 366 milliseconds.

Or we could extract the total seconds (including fractional seconds) by dividing by a 1 second `timedelta64`.

In [54]:
seconds = dt / np.timedelta64(1, 's')
print(seconds)

25620.366


Numpy has a special case for datetime errors called `NaT` (*not a time*). This is a variant of `NaN` (*not a number*).

In [55]:
np.datetime64("nat")
np.timedelta64("nat")

numpy.timedelta64('NaT')

In addition, Numpy stores a special case for when there is an error or the output isn't a valid datetime. This `NaT` value is good to know about because it usually means something has gone wrong in your system!

## Arithmetic with `np.datetime64`

Operations between a `datetime64` and a `timedelta64` **return** a `datetime64`.

In [56]:
ndt1 = np.datetime64("2018-10-03 13:00:00")
ndt2 = np.datetime64("2019-10-04 12:46:00")
delta1 = np.timedelta64(14, "D")

print(ndt1 - delta1)
print(ndt2 + delta1)

2018-09-19T13:00:00
2019-10-18T12:46:00


Arithmetic using Numpy objects is very similar to using Python `datetime` objects. The same rules apply to what we saw in our previous lesson. Just keep in mind whether your desired outcome is a **datetime** or a **timedelta** to check you are using the correct operation. Some arithmetic will return a time difference and others give you a fixed point in time. Arithmetic between a `datetime64` and a `timedelta64` will return a new `datetime64`.

**Subtracting** a `datetime64` from another `datetime64` will return the elapsed time between these two objects as a `timedelta64` object.

In [57]:
ndt2 - ndt1

numpy.timedelta64(31621560,'s')

You **cannot** add two `datetime64` objects.

Always use **time unit** codes to ensure your arithmetic changes a `datetime64` correctly.

In [58]:
ndt1 + np.timedelta64(7, "D") # Add 7 days to a datetime

numpy.datetime64('2018-10-10T13:00:00')

Since **Months** and **Years** have variable lengths - you cannot convert a `timedelta64` in either unit into days units. This is to ensure **consistency** in Numpy.

In [59]:
d1 = np.datetime64('2026-02')          # month precision: <M8[M]
d2 = d1 + np.timedelta64(1, 'M')       # 2026-03

(d2.astype('datetime64[D]') - d1.astype('datetime64[D]')) / np.timedelta64(1, 'D')
# 28.0

28.0

Understanding the time unit codes are the key to mastering `datetime64` in Numpy. We previously mentioned that there is a subtle distinction when converting some `timedelta64` objects with variable length into other units. Specifically, some units such as month and year aren't always the same size. Therefore, Numpy will prohibit you to convert a 1 month interval into days by default because we can't know whether this should be 28 days or 31 days. Similarly, a 1 year interval could be 365 days or 366 days. These safety checks only exist **when creating a new `np.timedelta64` object** and **not** when using `np.astype()`. The `astype()` behaviour will use default values in conversion so the result here would be 30. This _might_ be correct, but not always because months aren't always 30 days. Keep this in mind when converting between `timedelta64` units.

## Using `datetime64` for working days
Another useful feature of `np.datetime64` is built in functionality for detecting and handling **business days**. 

The `is_busday()` function return `True` if a given `np.datetime64` is in the Monday-Friday working week.

In [60]:
np.is_busday(np.datetime64("2020-08-12"))

True

`busday_count()` will return the number of business days between two dates.

In [61]:
ndt1 = np.datetime64("2020-09-20")
ndt2 = np.datetime64("2020-12-15")

np.busday_count(ndt1, ndt2)

61

Numpy includes some built in functions for handling business days and working weeks inside `np.datetime64`. While this will not always be useful, the time spent in understanding these functions is far less than the time reimplementing such functionality if you didn't know this existed. For example, this can be useful if you need to calculate shipping estimates for logistics pipelines or calculate productivity hours in given weeks. 

The `is_busday()` function is the simplest utility which only returns if a given date is a business day or not. The `busday_count()` functionality will return the number of business days between two dates. This is equivalent to summing all the days in this interval for which `is_busday()` returns `True`. Remember, the end date is always excluded. 

The `busday_offset()` function will calculate an offset of `k` business days from a given date.

In [62]:
ndt1 = np.datetime64("2020-08-12")
k=2
np.busday_offset(ndt1, k)

numpy.datetime64('2020-08-14')

Hence if you ask for an offset of 1 from a Friday, then the function will return the following Monday. 

This function assumes the input date is **also** a business day. If this is not true then the input day is **adjusted** using the `roll` argument.
- `roll='forward'`: adjust to first valid day in the future.
- `roll='backward'`: adjust to first valid day in the past.
- `roll='nat'`: return Not A Number.
- `roll='raise'`: raise exception instead of adjusting (default).

In [63]:
ndt2 = np.datetime64("2020-08-08") # Saturday 8th August 2020
np.busday_offset(ndt2, k, roll="forward")

numpy.datetime64('2020-08-12')

Here, we ask for 2 business days from `ndt2` but this object represents a Saturday. By specifying `roll=forward`, the date is adjusted to the following Monday and only then is the offset applied. We can also specify which days are considered business days using the `weekmask` argument. 

## Interacting with Pandas

* Numpy and Pandas can be used together when handling datetime.
* Pandas handles datetimes using the `Timestamp` class.

Create a `pandas.Timestamp` class from a `np.datetime64` object using `pd.to_datetime()` - [documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html)

In [64]:
import pandas as pd

timestamp = pd.to_datetime(ndt_now)
timestamp

Timestamp('2026-08-16 17:44:45.455570')

Pandas `Timestamp` works similarly to Python `datetime`. Use dot notation to access attributes.

In [65]:
timestamp.day

16

Finally, we will briefly discuss how Numpy interacts with the Pandas library for datetime objects. Pandas is the another important library for data processing and understanding both Numpy and Pandas can improve your data science projects immensely! 

Both libraries have dedicated classes for handling datetime objects and the Pandas variant is referred to as a `Timestamp`. You can create a Pandas `Timestamp` using the `pd.to_datetime()` function. Inside a `Timestamp`, pandas will still use efficient Numpy `datetime64` objects to store data. The advantage of using a `Timestamp` object is easy interaction with other Pandas functions which are powerful tools for handling time series data. In addition, the `Timestamp` class has many features from the Python `datetime` class - including dot notation access for attributes and time zones. We cover datetime handling in Pandas in more detail in other lessons within this module.